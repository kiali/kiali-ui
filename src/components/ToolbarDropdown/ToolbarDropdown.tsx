import * as React from 'react';
import { DropdownButton, MenuItem } from 'patternfly-react';
import PropTypes from 'prop-types';

type ToolbarDropdownProps = {
  disabled: boolean;
  nameDropdown?: string;
  initialValue: number | string;
  initialLabel: string | undefined;
  onClick: PropTypes.func;
  options: PropTypes.Map<string | number, string>;
};

type ToolbarDropdownState = {
  currentValue: number | string;
  currentName: string | undefined;
};

export class ToolbarDropdown extends React.Component<ToolbarDropdownProps, ToolbarDropdownState> {
  constructor(props: ToolbarDropdownProps) {
    super(props);
    this.state = {
      currentValue: props.initialValue,
      currentName: props.initialLabel
    };
  }

  onKeyChanged = (key: any) => {
    this.setState({ currentValue: key[0], currentName: key[1] });
    this.props.onClick(key[0]);
  };

  render() {
    return (
      <div className="form-group">
        {this.props.nameDropdown && <label style={{ paddingRight: '0.5em' }}>{this.props.nameDropdown}:</label>}
        <DropdownButton title={this.state.currentName} onSelect={this.onKeyChanged}>
          {this.props.options.map(r => (
            <MenuItem key={r[0]} active={r[0] === this.state.currentValue} eventKey={r}>
              {r[1]}
            </MenuItem>
          ))}
        </DropdownButton>
      </div>
    );
  }
}
