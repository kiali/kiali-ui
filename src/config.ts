import deepFreeze from 'deep-freeze';

export const config = () => {
  return deepFreeze({
    version: '0.1',
    toolbar: {
      defaultDuration: 60,
      intervalDuration: {
        60: 'Last minute',
        300: 'Last 5 minutes',
        600: 'Last 10 minutes',
        1800: 'Last 30 minutes',
        3600: 'Last hour',
        10800: 'Last 3 hours',
        21600: 'Last 6 hours',
        43200: 'Last 12 hours',
        86400: 'Last day',
        604800: 'Last 7 days',
        2592000: 'Last 30 days'
      },
      defaultPollInterval: 15000,
      pollInterval: {
        0: 'Pause',
        5000: '5 seconds',
        10000: '10 seconds',
        15000: '15 seconds',
        30000: '30 seconds',
        60000: '1 minute',
        300000: '5 minutes'
      },
      graphLayouts: {
        breadthfirst: 'Breadthfirst',
        cola: 'Cola',
        cose: 'Cose',
        dagre: 'Dagre',
        klay: 'Klay'
      }
    }
  });
};

export const KEY_CODES = { TAB_KEY: 9, ENTER_KEY: 13, ESCAPE_KEY: 27 };

export const socialLinks = [
  {
    url: 'https://github.com/kiali',
    icon: { type: 'fa', name: 'github' },
    label: 'Github'
  },
  {
    url: 'https://www.youtube.com/channel/UCcm2NzDN_UCZKk2yYmOpc5w',
    icon: { type: 'fa', name: 'youtube' },
    label: 'Youtube'
  },
  {
    url: 'https://twitter.com/KialiProject',
    icon: { type: 'fa', name: 'twitter' },
    label: 'Twitter'
  }
];
