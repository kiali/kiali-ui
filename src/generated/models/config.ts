/* tslint:disable */
/* eslint-disable */
/**
 * Kiali
 * Kiali project, observability for the Istio service mesh
 *
 * The version of the OpenAPI document: _
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { Elements } from './elements';

/**
 *
 * @export
 * @interface Config
 */
export interface Config {
  /**
   *
   * @type {number}
   * @memberof Config
   */
  duration?: number;
  /**
   *
   * @type {Elements}
   * @memberof Config
   */
  elements?: Elements;
  /**
   *
   * @type {string}
   * @memberof Config
   */
  graphType?: string;
  /**
   *
   * @type {number}
   * @memberof Config
   */
  timestamp?: number;
}
