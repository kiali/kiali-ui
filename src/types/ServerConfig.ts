import { DurationInSeconds } from './Common';

export type IstioLabelKey = 'appLabelName' | 'versionLabelName';

// 3scale public config, typically to check if addon/extension is enabled
interface ThreeScaleConfig {
  enabled: boolean;
}
interface iter8Config {
  enabled: boolean;
}
// Kiali addons/extensions specific
interface Extensions {
  threescale: ThreeScaleConfig;
  iter8: iter8Config;
}

export interface ServerConfig {
  extensions?: Extensions;
  installationTag?: string;
  istioIdentityDomain: string;
  istioNamespace: string;
  istioComponentNamespaces?: Map<string, string>;
  istioLabels: { [key in IstioLabelKey]: string };
  prometheus: {
    globalScrapeInterval?: DurationInSeconds;
    storageTsdbRetention?: DurationInSeconds;
  };
  istioTelemetryV2: boolean;
}
