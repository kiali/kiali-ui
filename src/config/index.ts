// Configuration

import { config } from './config';

// Icons
import { icons } from './icons';

// Logos
import { kialiLogo } from './logos';

// Paths
import { Paths } from './Paths';

// Jaeger Query
import { JAEGER_QUERY } from './jaegerQuery';

export { config, Paths, icons, kialiLogo, JAEGER_QUERY };
