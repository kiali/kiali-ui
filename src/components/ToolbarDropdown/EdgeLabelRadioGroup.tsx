import * as React from 'react';
import PropTypes from 'prop-types';
import { style } from 'typestyle';
import * as _ from 'lodash';

import { GraphParamsType } from '../../types/Graph';
import { EdgeLabelMode } from '../../types/GraphFilter';

const EDGE_LABEL_MODES = EdgeLabelMode.getValues().reduce((map, edgeLabelMode) => {
  map[edgeLabelMode] = _.capitalize(_.startCase(edgeLabelMode));
  return map;
}, {});

type EdgeLabelRadioGroupProps = {
  graphParams: GraphParamsType;
  onEdgeChanged: PropTypes.func;
};

export class EdgeLabelRadioGroup extends React.Component<EdgeLabelRadioGroupProps> {
  constructor(props: EdgeLabelRadioGroupProps) {
    super(props);
  }

  onEdgeChanged = (event: any) => {
    const selectedEdge = event.target.value;
    this.setState({
      edgeValue: selectedEdge
    });
    this.props.onEdgeChanged(event);
  };

  render() {
    const radioButtonStyle = style({ marginLeft: 5 });

    const edgeItems = Object.keys(EDGE_LABEL_MODES).map((edgeLabelModeKey: any) => (
      <div key={edgeLabelModeKey}>
        <label className={radioButtonStyle}>
          <input
            type="radio"
            name="edge-label-radio-group"
            value={edgeLabelModeKey}
            checked={edgeLabelModeKey === this.props.graphParams.edgeLabelMode}
            onChange={this.onEdgeChanged}
          />
          <span className={radioButtonStyle}>{EDGE_LABEL_MODES[edgeLabelModeKey]}</span>
        </label>
      </div>
    ));

    return (
      <>
        <div>
          <label>Edge Labels:</label>
        </div>
        {edgeItems}
      </>
    );
  }
}

export default EdgeLabelRadioGroup;
