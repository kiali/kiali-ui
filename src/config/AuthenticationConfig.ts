import { AuthConfig, AuthStrategy } from '../types/Auth';

const authenticationConfig: AuthConfig = {
  strategy: AuthStrategy.login,
  secretMissing: false
};

// Returns true if authentication strategy is either 'openshift' or 'openid'
export const isAuthStrategyOAuth = () =>
  authenticationConfig.strategy === AuthStrategy.openshift || authenticationConfig.strategy === AuthStrategy.openid;

export default authenticationConfig;
