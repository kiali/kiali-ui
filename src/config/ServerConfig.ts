import deepFreeze from 'deep-freeze';
import { store } from '../store/ConfigStore';
import { KialiAppState, ServerConfig } from '../store/Store';
import { PersistPartial } from 'redux-persist';
import { DurationInSeconds } from '../types/Common';
import { forOwn, pickBy } from 'lodash';

// It's not great to access the store directly for convenience but the alternative is
// a huge code ripple just to access some server config. better to just have this one utility.
export const serverConfig = (): ServerConfig => {
  const actualState = store.getState() || ({} as KialiAppState & PersistPartial);
  return deepFreeze(actualState.serverConfig);
};

// getValidDurations returns a new object with only the durations <= retention
export const getValidDurations = (
  durations: { [key: string]: string },
  retention?: DurationInSeconds
): { [key: string]: string } => {
  const validDurations = pickBy(durations, (_, key) => {
    return !retention || Number(key) <= retention;
  });
  return validDurations;
};

// getValidDuration returns duration if it is a valid property key in durations, otherwise it return the first property
// key in durations.  It is assumed that durations has at least one property.
export const getValidDuration = (
  durations: { [key: string]: string },
  duration: DurationInSeconds
): DurationInSeconds => {
  let validDuration = 0;
  forOwn(durations, (value, key) => {
    const d = Number(key);
    if (d === duration) {
      validDuration = d;
    } else if (!validDuration) {
      validDuration = d;
    }
  });
  return validDuration;
};
