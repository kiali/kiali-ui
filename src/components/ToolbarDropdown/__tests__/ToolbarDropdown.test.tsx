import * as React from 'react';
import { mount, shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import ToolbarDropdown from '../ToolbarDropdown';
import { config } from '../../../config';
import { serverConfig } from '../../../config/ServerConfig';

const optionsChanged = jest.fn();

const data = [
  {
    id: 'graph_filter_interval_duration',
    default: config.toolbar.defaultDuration,
    options: serverConfig.durations
  },
  {
    id: 'metrics_filter_poll_interval',
    default: config.toolbar.defaultPollInterval,
    options: config.toolbar.pollInterval
  },
  { id: 'graph_filter_layouts', default: 'cola', options: config.toolbar.graphLayouts }
];

describe('ToolbarDropdown', () => {
  it('Render correctly all dropdowns', () => {
    data.forEach(dropdownType => {
      const wrapper = shallow(
        <ToolbarDropdown
          id={dropdownType.id}
          disabled={false}
          handleSelect={jest.fn()}
          nameDropdown={dropdownType.id}
          initialValue={dropdownType.default}
          initialLabel={dropdownType.options[dropdownType.default]}
          options={dropdownType.options}
        />
      );
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });

  it('Render dropdowns correctly with controlled values and labels', () => {
    data.forEach(dropdownType => {
      const wrapper = shallow(
        <ToolbarDropdown
          id={dropdownType.id}
          disabled={false}
          handleSelect={jest.fn()}
          nameDropdown={dropdownType.id}
          value={dropdownType.default}
          label={dropdownType.options[dropdownType.default]}
          options={dropdownType.options}
        />
      );
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });

  it('changes trigger parent callback', () => {
    const wrapper = mount(
      <ToolbarDropdown
        id={'graph_filter_interval_duration'}
        disabled={false}
        handleSelect={optionsChanged}
        nameDropdown={'Duration'}
        initialValue={config.toolbar.defaultDuration}
        initialLabel={serverConfig.durations[config.toolbar.defaultDuration]}
        options={serverConfig.durations}
      />
    );
    const elt = wrapper
      .find('#graph_filter_interval_duration')
      .find('SafeAnchor')
      .first();
    elt.simulate('click');
    expect(optionsChanged).toHaveBeenCalledTimes(1);
  });

  it('Check properties', () => {
    const idElement = 'MyToolbarDropdown';
    const nameDropdownElt = 'NameDropdownElt';
    const initialValue = 'InitialValue';
    const initialLabel = 'InitialLabel';
    const options = {
      60: 'Last minute',
      300: 'Last 5 minutes',
      600: 'Last 10 minutes',
      1800: 'Last 30 minutes',
      3600: 'Last hour',
      10800: 'Last 3 hours',
      21600: 'Last 6 hours',
      43200: 'Last 12 hours',
      86400: 'Last day',
      604800: 'Last 7 days',
      2592000: 'last 30 days'
    };
    const wrapper = shallow(
      <ToolbarDropdown
        id={'MyToolbarDropdown'}
        disabled={false}
        handleSelect={optionsChanged}
        nameDropdown={nameDropdownElt}
        initialValue={initialValue}
        initialLabel={initialLabel}
        options={options}
      />
    );
    const elt = wrapper.find('#' + idElement);
    expect(elt.prop('title')).toEqual(initialLabel);
    expect(elt.prop('id')).toEqual(idElement);
    expect(elt.children().length).toEqual(Object.keys(options).length);
  });
});
