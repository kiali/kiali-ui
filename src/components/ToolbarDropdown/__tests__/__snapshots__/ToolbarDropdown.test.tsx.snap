// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolbarDropdown Render correctly all dropdowns 1`] = `
<Fragment>
  <label
    style={
      Object {
        "paddingRight": "0.5em",
      }
    }
  >
    graph_filter_interval_duration
  </label>
  <DropdownButton
    disabled={false}
    id="graph_filter_interval_duration"
    onSelect={[Function]}
    onToggle={[Function]}
    title="Last 1m"
  >
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="60"
      header={false}
      key="60"
    >
      Last 1m
    </MenuItem>
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="300"
      header={false}
      key="300"
    >
      Last 5m
    </MenuItem>
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="600"
      header={false}
      key="600"
    >
      Last 10m
    </MenuItem>
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="1800"
      header={false}
      key="1800"
    >
      Last 30m
    </MenuItem>
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="3600"
      header={false}
      key="3600"
    >
      Last 1h
    </MenuItem>
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="10800"
      header={false}
      key="10800"
    >
      Last 3h
    </MenuItem>
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="21600"
      header={false}
      key="21600"
    >
      Last 6h
    </MenuItem>
  </DropdownButton>
</Fragment>
`;

exports[`ToolbarDropdown Render correctly all dropdowns 2`] = `
<Fragment>
  <label
    style={
      Object {
        "paddingRight": "0.5em",
      }
    }
  >
    metrics_filter_poll_interval
  </label>
  <DropdownButton
    disabled={false}
    id="metrics_filter_poll_interval"
    onSelect={[Function]}
    onToggle={[Function]}
    title="Every 15s"
  >
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="0"
      header={false}
      key="0"
    >
      Pause
    </MenuItem>
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="10000"
      header={false}
      key="10000"
    >
      Every 10s
    </MenuItem>
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="15000"
      header={false}
      key="15000"
    >
      Every 15s
    </MenuItem>
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="30000"
      header={false}
      key="30000"
    >
      Every 30s
    </MenuItem>
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="60000"
      header={false}
      key="60000"
    >
      Every 1m
    </MenuItem>
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="300000"
      header={false}
      key="300000"
    >
      Every 5m
    </MenuItem>
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="900000"
      header={false}
      key="900000"
    >
      Every 15m
    </MenuItem>
  </DropdownButton>
</Fragment>
`;

exports[`ToolbarDropdown Render correctly all dropdowns 3`] = `
<Fragment>
  <label
    style={
      Object {
        "paddingRight": "0.5em",
      }
    }
  >
    graph_filter_layouts
  </label>
  <DropdownButton
    disabled={false}
    id="graph_filter_layouts"
    onSelect={[Function]}
    onToggle={[Function]}
    title="Cola"
  >
    <MenuItem
      active={true}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="cola"
      header={false}
      key="cola"
    >
      Cola
    </MenuItem>
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="cose-bilkent"
      header={false}
      key="cose-bilkent"
    >
      Cose
    </MenuItem>
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="dagre"
      header={false}
      key="dagre"
    >
      Dagre
    </MenuItem>
  </DropdownButton>
</Fragment>
`;

exports[`ToolbarDropdown Render dropdowns correctly with controlled values and labels 1`] = `
<Fragment>
  <label
    style={
      Object {
        "paddingRight": "0.5em",
      }
    }
  >
    graph_filter_interval_duration
  </label>
  <DropdownButton
    disabled={false}
    id="graph_filter_interval_duration"
    onSelect={[Function]}
    onToggle={[Function]}
    title="Last 1m"
  >
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="60"
      header={false}
      key="60"
    >
      Last 1m
    </MenuItem>
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="300"
      header={false}
      key="300"
    >
      Last 5m
    </MenuItem>
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="600"
      header={false}
      key="600"
    >
      Last 10m
    </MenuItem>
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="1800"
      header={false}
      key="1800"
    >
      Last 30m
    </MenuItem>
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="3600"
      header={false}
      key="3600"
    >
      Last 1h
    </MenuItem>
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="10800"
      header={false}
      key="10800"
    >
      Last 3h
    </MenuItem>
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="21600"
      header={false}
      key="21600"
    >
      Last 6h
    </MenuItem>
  </DropdownButton>
</Fragment>
`;

exports[`ToolbarDropdown Render dropdowns correctly with controlled values and labels 2`] = `
<Fragment>
  <label
    style={
      Object {
        "paddingRight": "0.5em",
      }
    }
  >
    metrics_filter_poll_interval
  </label>
  <DropdownButton
    disabled={false}
    id="metrics_filter_poll_interval"
    onSelect={[Function]}
    onToggle={[Function]}
    title="Every 15s"
  >
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="0"
      header={false}
      key="0"
    >
      Pause
    </MenuItem>
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="10000"
      header={false}
      key="10000"
    >
      Every 10s
    </MenuItem>
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="15000"
      header={false}
      key="15000"
    >
      Every 15s
    </MenuItem>
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="30000"
      header={false}
      key="30000"
    >
      Every 30s
    </MenuItem>
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="60000"
      header={false}
      key="60000"
    >
      Every 1m
    </MenuItem>
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="300000"
      header={false}
      key="300000"
    >
      Every 5m
    </MenuItem>
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="900000"
      header={false}
      key="900000"
    >
      Every 15m
    </MenuItem>
  </DropdownButton>
</Fragment>
`;

exports[`ToolbarDropdown Render dropdowns correctly with controlled values and labels 3`] = `
<Fragment>
  <label
    style={
      Object {
        "paddingRight": "0.5em",
      }
    }
  >
    graph_filter_layouts
  </label>
  <DropdownButton
    disabled={false}
    id="graph_filter_layouts"
    onSelect={[Function]}
    onToggle={[Function]}
    title="Cola"
  >
    <MenuItem
      active={true}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="cola"
      header={false}
      key="cola"
    >
      Cola
    </MenuItem>
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="cose-bilkent"
      header={false}
      key="cose-bilkent"
    >
      Cose
    </MenuItem>
    <MenuItem
      active={false}
      bsClass="dropdown"
      disabled={false}
      divider={false}
      eventKey="dagre"
      header={false}
      key="dagre"
    >
      Dagre
    </MenuItem>
  </DropdownButton>
</Fragment>
`;
